Remove Frame::setOpener

achristensen07 opened the pull request:
#### b87d18215a73f63b17fc05ca68ad57d8a04646e4
<pre>
Remove Frame::setOpener
<a href="https://bugs.webkit.org/show_bug.cgi?id=278592">https://bugs.webkit.org/show_bug.cgi?id=278592</a>
<a href="https://rdar.apple.com/134596394">rdar://134596394</a>

Reviewed by Charlie Wolfe.

This PR untangles the different cases where we need to set
the opener of a frame.  Most commonly we only want to remove
the opener.  A frame can usually only get an opener if it is
created with one, which I moved to the UI process in
274396@main but it can also be given an opener later if
window.open is called with an existing frame as the target.
We also need a setter upon Page construction in
WebKitLegacy because the delegate callback doesn&apos;t have a
configuration object and when switching between a LocalFrame
and RemoteFrame.

This is step 1 of turning FrameLoaderClient::dispatchCreatePage
inside out and making it send the request and other parameters
for the UI process to decide to make a page in whatever process
it decides to.  This is needed for implementing window.open
with noopener with site isolation enabled.

In order for this new way of setting the opener to work,
users of WKPageSetPageUIClient need to be updated to v6
to get the createNewPage callback with a configuration where
we can preserve the opener info like we do with WKUIDelegate.
I&apos;ve gone through the users of WKPageSetPageUIClient and made
sure this won&apos;t break any current use.  There is one current
use and it always returns nil, so there is no compatibility
concern there.

WebCore::createWindow had an issue that had been covered up for
many years.  It takes two LocalFrame parameters, but when it is
called from LocalDOMWindow::createWindow the parameter order was
switched, and we used to just call Frame::setOpener to cover it
up and make sure the frame had the correct opener when we were
done.  With this new flow of giving the Frame the correct opener
in its constructor then never changing it unless we call open
with a named existing frame, we need to use the correct frame as
the opener.  Since we were already using the opener as the lookup
frame, I simplified it to only take one frame as a parameter,
the correct opener frame.

* Source/WebCore/inspector/InspectorFrontendClientLocal.cpp:
(WebCore::InspectorFrontendClientLocal::openURLExternally):
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::~FrameLoader):
(WebCore::FrameLoader::continueLoadAfterNewWindowPolicy):
(WebCore::FrameLoader::switchBrowsingContextsGroup):
* Source/WebCore/page/Frame.cpp:
(WebCore::Frame::Frame):
(WebCore::Frame::takeWindowProxyAndOpenerFrom):
(WebCore::Frame::disownOpener):
(WebCore::Frame::setOpenerForWebKitLegacy):
(WebCore::Frame::detachFromAllOpenedFrames):
(WebCore::Frame::takeWindowProxyFrom): Deleted.
(WebCore::Frame::setOpener): Deleted.
(WebCore::Frame::openedFrames): Deleted.
* Source/WebCore/page/Frame.h:
(WebCore::Frame::opener):
* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::disownOpener):
(WebCore::LocalDOMWindow::createWindow):
* Source/WebCore/page/RemoteDOMWindow.cpp:
(WebCore::RemoteDOMWindow::setOpener): Deleted.
* Source/WebCore/page/RemoteDOMWindow.h:
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundleFrame.cpp:
(WKBundleFrameClearOpener):
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::loadDidCommitInAnotherProcess):
(WebKit::WebFrame::commitProvisionalFrame):
* Source/WebKitLegacy/mac/WebCoreSupport/WebFrameLoaderClient.mm:
(WebFrameLoaderClient::dispatchCreatePage):
* Source/WebKitLegacy/mac/WebView/WebFrame.mm:
(-[WebFrame _clearOpener]):
* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebView _commonInitializationWithFrameName:groupName:]):
(-[WebView initSimpleHTMLDocumentWithStyle:frame:preferences:groupName:]):

Canonical link: <a href="https://commits.webkit.org/282987@main">https://commits.webkit.org/282987@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b642607f4e4189654c7866a37711e60c98d2471c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/64760 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/44125 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/17356 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/68784 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/15367 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/66877 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/51906 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/15645 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/52062 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/10602 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/67826 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/40819 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/56031 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/32681 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/37484 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/13406 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/14243 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/59412 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/13736 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/70490 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/8705 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/13230 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/59388 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/8739 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/56115 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/59587 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/7186 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/858 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9833 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/39937 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/41016 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/42197 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/40758 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

djg commented on Source/WebCore/inspector/InspectorFrontendClientLocal.cpp:
```diff
@@ -290,8 +290,9 @@ void InspectorFrontendClientLocal::openURLExternally(const String& url)
 
-    frame->setOpener(mainFrame.ptr());
+    ASSERT(frame->opener() == mainFrame.ptr());
     frame->page()->setOpenedByDOM();
+    frame->page()->setOpenedByDOMWithOpener(true);
```
This feels like an odd API, in that, `setOpenedByDOMWithOpener` sounds like a superset of `setOpenedByDOM` but `setOpenedByDOMWithOpener` doesn't set `m_openedByDOM`. Are `m_openedByDOM` and `m_openedByDOMWithOpener` really a tri-state?

achristensen07 replied:
It is odd, but they are distinct, and also distinct from just having an opener.  I think I'm on my way to removing setOpenedByDOMWithOpener, but that can't be done in this PR.

charliewolfe commented on Source/WebKitLegacy/mac/WebCoreSupport/WebChromeClient.mm:
```diff
@@ -304,8 +304,11 @@ - (void)setIsSelected:(BOOL)isSelected;
-    if (newPage && !features.wantsNoOpener())
+    if (newPage && !features.wantsNoOpener()) {
         m_webView.page->protectedStorageNamespaceProvider()->cloneSessionStorageNamespaceForPage(*m_webView.page, *newPage);
+        if (newWebView)
```
nit - this if isn't needed

charliewolfe reviewed (commented):
The asserts you added in `LocalDOMWindow::createWindow` are being hit by EWS

achristensen07 committed:
Remove Frame::setOpener
https://bugs.webkit.org/show_bug.cgi?id=278592
rdar://134596394

Reviewed by Charlie Wolfe.

This PR untangles the different cases where we need to set
the opener of a frame.  Most commonly we only want to remove
the opener.  A frame can usually only get an opener if it is
created with one, which I moved to the UI process in
274396@main but it can also be given an opener later if
window.open is called with an existing frame as the target.
We also need a setter upon Page construction in
WebKitLegacy because the delegate callback doesn't have a
configuration object and when switching between a LocalFrame
and RemoteFrame.

This is step 1 of turning FrameLoaderClient::dispatchCreatePage
inside out and making it send the request and other parameters
for the UI process to decide to make a page in whatever process
it decides to.  This is needed for implementing window.open
with noopener with site isolation enabled.

In order for this new way of setting the opener to work,
users of WKPageSetPageUIClient need to be updated to v6
to get the createNewPage callback with a configuration where
we can preserve the opener info like we do with WKUIDelegate.
I've gone through the users of WKPageSetPageUIClient and made
sure this won't break any current use.  There is one current
use and it always returns nil, so there is no compatibility
concern there.

WebCore::createWindow had an issue that had been covered up for
many years.  It takes two LocalFrame parameters, but when it is
called from LocalDOMWindow::createWindow the parameter order was
switched, and we used to just call Frame::setOpener to cover it
up and make sure the frame had the correct opener when we were
done.  With this new flow of giving the Frame the correct opener
in its constructor then never changing it unless we call open
with a named existing frame, we need to use the correct frame as
the opener.  Since we were already using the opener as the lookup
frame, I simplified it to only take one frame as a parameter,
the correct opener frame.

* Source/WebCore/inspector/InspectorFrontendClientLocal.cpp:
(WebCore::InspectorFrontendClientLocal::openURLExternally):
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::~FrameLoader):
(WebCore::FrameLoader::continueLoadAfterNewWindowPolicy):
(WebCore::FrameLoader::switchBrowsingContextsGroup):
* Source/WebCore/page/Frame.cpp:
(WebCore::Frame::Frame):
(WebCore::Frame::takeWindowProxyAndOpenerFrom):
(WebCore::Frame::disownOpener):
(WebCore::Frame::setOpenerForWebKitLegacy):
(WebCore::Frame::detachFromAllOpenedFrames):
(WebCore::Frame::takeWindowProxyFrom): Deleted.
(WebCore::Frame::setOpener): Deleted.
(WebCore::Frame::openedFrames): Deleted.
* Source/WebCore/page/Frame.h:
(WebCore::Frame::opener):
* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::disownOpener):
(WebCore::LocalDOMWindow::createWindow):
* Source/WebCore/page/RemoteDOMWindow.cpp:
(WebCore::RemoteDOMWindow::setOpener): Deleted.
* Source/WebCore/page/RemoteDOMWindow.h:
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundleFrame.cpp:
(WKBundleFrameClearOpener):
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::loadDidCommitInAnotherProcess):
(WebKit::WebFrame::commitProvisionalFrame):
* Source/WebKitLegacy/mac/WebCoreSupport/WebFrameLoaderClient.mm:
(WebFrameLoaderClient::dispatchCreatePage):
* Source/WebKitLegacy/mac/WebView/WebFrame.mm:
(-[WebFrame _clearOpener]):
* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebView _commonInitializationWithFrameName:groupName:]):
(-[WebView initSimpleHTMLDocumentWithStyle:frame:preferences:groupName:]):

Canonical link: https://commits.webkit.org/282987@main
